### ISSUE TITLE
How can I find the process of calling a function in benchdnn or gtest(like convolution)?

### ISSUE THREAD

**Issue by knjwhn** [intel/mkl-dnn#562]
Hello, everyone
Here is my problem, I want to add some custom function and make it pass the test and use it in tensorflow or mxnet finally.  I make install the mkldnn in linux successful and I'm looking at source code in windows manually(/tests/benchdnn/benchdnn.cpp), unfortunately, I still can't find the process.  Anyone can help or any advices? Thanks a lot.

**Comment by rsdubtso** [comment]
Hi! Can you please elaborate? Is this related to this issue: https://github.com/apache/incubator-mxnet/issues/15419?

**Comment by knjwhn** [comment]
Yeah,I made that issue. And now I plan to add some custom function in mkldnn , and I'm trying to figure out the process of calling a function(like convolution or activation), Do you have some sugesstion? many thanks!

**Comment by emfomenk** [comment]
If the primary goal to experiment with a new feature that you want to add it might make sense to start with changing some simple example, for instance these few:
- [getting_started.cpp](https://github.com/intel/mkl-dnn/blob/master/examples/getting_started.cpp). If one removes the comments the example simply call ReLU and verifies the results.
- [memory_format_propagation.cpp](https://github.com/intel/mkl-dnn/blob/master/examples/memory_format_propagation.cpp). Again, most of the file are the comments. The example simplн executes convolution and subsequent pooling based on the layout from convolution.

[Benchdnn](https://github.com/intel/mkl-dnn/tree/master/tests/benchdnn) is a complex program that can check performance and correctness of DNNL primitives. In essence, benchdnn structure is:
- root directory -- services functions for benchdnn and its drivers
  - `benchdnn.cpp` -- main harness that dispatches drivers like tests for conv, tests for batch norm, etc... If you want to add a new primitive and the corresponding driver to benchdnn, probably that would be the only one file that you need to modify in the root directory.
- `$driver`/ -- directory for testing particular primitive, for instance conv.
  - `bench_$driver.cpp` -- files that parses command line arguments
  - `$driver.cpp` -- file that initializes and executes a primitive (see `doit()` function), and also check the correctness of the output (typically by calling `compare()` function).
  - `ref_$driver.cpp` -- reference implementation that is used to compare result with.

Benchdnn also has a [README.md](https://github.com/intel/mkl-dnn/blob/master/tests/benchdnn/README.md) file that explains the command line arguments and also shed some lite on the testing approach.

In short, the common testing methodology can be depicted as:
```
           start
             |
             V
    ref_input [nchw format]   --[reorder]-->   dnnl_input [testing format]
             |                                           |
             V                                           V
      ref computations                           dnnl computations
             |                                           |
             V                                           V
   ref_output [nchw format]                   dnnl_output [testing format]
             |                                           |
             |                                       [reorder]
             |                                           |
             V                                           V
         <compare>   <---------------------   dnnl_output [ref format]
             |
             V
          result
```

**Comment by knjwhn** [comment]
> If the primary goal to experiment with a new feature that you want to add it might make sense to start with changing some simple example, for instance these few:
> 
> * [getting_started.cpp](https://github.com/intel/mkl-dnn/blob/master/examples/getting_started.cpp). If one removes the comments the example simply call ReLU and verifies the results.
> * [memory_format_propagation.cpp](https://github.com/intel/mkl-dnn/blob/master/examples/memory_format_propagation.cpp). Again, most of the file are the comments. The example simplн executes convolution and subsequent pooling based on the layout from convolution.
> 
> [Benchdnn](https://github.com/intel/mkl-dnn/tree/master/tests/benchdnn) is a complex program that can check performance and correctness of DNNL primitives. In essence, benchdnn structure is:
> 
> * root directory -- services functions for benchdnn and its drivers
>   
>   * `benchdnn.cpp` -- main harness that dispatches drivers like tests for conv, tests for batch norm, etc... If you want to add a new primitive and the corresponding driver to benchdnn, probably that would be the only one file that you need to modify in the root directory.
> * `$driver`/ -- directory for testing particular primitive, for instance conv.
>   
>   * `bench_$driver.cpp` -- files that parses command line arguments
>   * `$driver.cpp` -- file that initializes and executes a primitive (see `doit()` function), and also check the correctness of the output (typically by calling `compare()` function).
>   * `ref_$driver.cpp` -- reference implementation that is used to compare result with.
> 
> Benchdnn also has a [README.md](https://github.com/intel/mkl-dnn/blob/master/tests/benchdnn/README.md) file that explains the command line arguments and also shed some lite on the testing approach.
> 
> In short, the common testing methodology can be depicted as:
> 
> ```
>            start
>              |
>              V
>     ref_input [nchw format]   --[reorder]-->   dnnl_input [testing format]
>              |                                           |
>              V                                           V
>       ref computations                           dnnl computations
>              |                                           |
>              V                                           V
>    ref_output [nchw format]                   dnnl_output [testing format]
>              |                                           |
>              |                                       [reorder]
>              |                                           |
>              V                                           V
>          <compare>   <---------------------   dnnl_output [ref format]
>              |
>              V
>           result
> ```

Thanks for your help! In my opinion, First I should find the process of calling a function, like How can I know the mkldnn do after I call the relu.execute(...) , Is there have some method to use like gdb or something?

**Comment by emfomenk** [comment]
You have the following options:
- `export MKLDNN_VERBOSE=1 DNNL_VERBOSE=1` to see the verbose output of the primitive execution (see this [section](https://intel.github.io/mkl-dnn/dev_guide_verbose.html))
- set breakpoint to some internal function (that you are interested in). All primitives should go through `dnnl_primitive_execute()` (v1.1 and after) or `mkldnn_primitive_execute()` (before v1.1).
- validate the result :)

**Comment by knjwhn** [comment]
Thank you! And may I ask how to use the breakpoint to find the internal function that are called? Sorry I am still a beginner :)

**Comment by emfomenk** [comment]
gdb is very powerful debugger. Probably, the simplest way to try something is to put a breakpoint of a file with line number (instead of setting breakpoint on internal function which would have an awfully long name due to nested namespaces and class + templates).
So assuming you run convolution on avx2 system you may want to set break points to:
- [`$b jit_avx2_1x1_convolution.cpp :39`](https://github.com/intel/mkl-dnn/blob/master/src/cpu/jit_avx2_1x1_convolution.cpp#L39)
- [`$b jit_avx2_convolution.cpp :46`](https://github.com/intel/mkl-dnn/blob/master/src/cpu/jit_avx2_convolution.cpp#L46)

It is also very important to build DNNL in debug mode: `cmake -DCMAKE_BUILD_TYPE=DEBUG ..`.

It is worth mentioning that most DNNL kernels are written using Just In Time code generation (using [xbyak](https://github.com/herumi/xbyak)) -- essentially, the library generates assembly code on the fly during run-time. This makes it very hard to debug: there is no obvious breakpoint you can set on, and even if you happen to be at the right place you will be able to see only assembly code...

**Comment by knjwhn** [comment]
Ok! Really really Thanks! I will try it.And hope you have a nice day!

**Comment by vpirogov** [comment]
Closing as the issue is resolved. Feel free to reopen or submit a new one if you have any further questions!
